keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2150 @@
+the pod spec>`. Ownership must be checked to detect accidental conflicts with
+persistent ResourceClaims or previous incarnations of the same ephemeral
+resource. This is the same approach that was chosen for [generic ephemeral
+volumes](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md).
+For a resource driver there is no difference.
+
+### Coordinating resource allocation through the scheduler
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. The downside is that pod scheduling is less flexible.
+
+For delayed allocation, a node is selected tentatively by the scheduler
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource
+            is allocated and unused,
+            uses delayed allocation, and
+            was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
+    }
+  } else if <all resources allocated> {
+    <schedule pod onto node>
+  } else if <some unallocated resource uses delayed allocation> {
+    <tell resource driver to allocate for the chosen node>
+  }
+}
+```
+
+### Resource allocation and usage flow
+
+The following steps shows how resource allocation works for a resource that
+gets defined inline in a Pod. Several of these steps may fail without changing
+the system state. They then must be retried until they succeed or something
+else changes in the system, like for example deleting objects.
+
+* **user** creates Pod with inline ResourceClaimTemplate
+* **resource claim controller** checks ResourceClaimTemplate and ResourceClass,
+  then creates ResourceClaim with Pod as owner
+* if *immediate allocation*:
+  * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+  * **resource driver** finishes allocation, sets `Allocation` -> claim ready for use by any pod
+* if *pod is pending*:
+  * **scheduler** filters nodes
+  * if *delayed allocation and resource not allocated yet*:
+    * if *at least one node fits pod*:
+      * **scheduler** picks one node, sets `SuitableNodes`, `SelectedNode` and `SelectedUser`
```

Comment:
does the scheduler also pick a ResourceClaim or the RC already belongs to the Pod?

Is `SelectedUser` the Pod?

Reply:
> does the scheduler also pick a ResourceClaim or the RC already belongs to the Pod?

The Pod spec defines which ResourceClaims are used by a Pod, via the `PodSpec.ResourceClaims` list. The scheduler does not need pick one.

> Is SelectedUser the Pod?

Yes. The API comment for the field explains that further. I'm struggling a bit here with what should come first: the design details that explain why the API is what it is, or the API :cry: 

Would you prefer to have the API section first or does that just raise other questions because the existence of a field hasn't been motivated by giving the necessary context?

Reply:
->

> **scheduler** picks one node, sets `SuitableNodes=<all nodes that fit the pod>`, `SelectedNode=<the chosen node>` and `SelectedUser=<the pod being scheduled>`